### PR TITLE
Questions component rework

### PR DESCRIPTION
**Pull request by vrackovic**
- Added form submit functionality
- Works with Post and Put requests

### CODE REVIEW

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 1]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
+  @observable answers = [{id: 1, description: '', isCorrect: ''}];
+
+  @observable selectedCategory = '';
+
+  @observable
+  questionTypes = ['FREE_FORM', 'MULTIPLE_CHOICE', 'CODING_CHALLENGE'];
+
+  @action.bound
+  incrementAnswers() {
+    const lastElementId = this.answers.slice(-1)[0].id;
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: This isn't really readable. How about:

```js
// reverse shouldn't mutate answers
const [{ id: lastId }] = reverse(this.answers);
```

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 2]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
+
+  @action.bound
+  decrementAnswers(answerId) {
+    if (this.answers.length > 1) {
+      this.answers = this.answers.filter(answer => answer.id !== answerId);
+    }
+  }
+
+  @action.bound
+  getQuestionById(id) {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Rename to `setSingleQuestionById`.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 3]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
+    this.questions.map(question => {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Use `Array#find`.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 4]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
-      .then(
-        questions =>
-          (this.questions = questions.map(this.prepareQuestionForView))
-      );
+      .then(questions => (this.questions = questions));
+  }
+
+  @action.bound
+  submitQuestion(question) {
+    if (question.id) {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: No need to duplicate:

```js
const submit = question.id ? questionApi.updateQuestion : questionApi.createNewQuestion;

return submit(question).then(() => this.fetchQuestions());
```

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 5]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
   get questionTable() {
-    return util.tableTransformer([...this.questions]);
+    if (!this.selectedCategory) {
+      return util.tableTransformer([
+        ...this.questions.map(util.prepareQuestionForView)
+      ]);
+    }
+    return util.tableTransformer([
+      ...this.questions
+        .filter(question => question.category === this.selectedCategory)
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: No need to duplicate:

```js
const getQuestionsForCategory = this.selectedCategory
  ? question => question.category === this.selectedCategory)
  : () => true;

return util.tableTransformer([
  ...this.questions.map(util.prepareQuestionForView)
    .filter(getQuestionsForCategory)
]);
```

File: web/src/main/frontend/src/components/Question.js [hunk 6]
```diff
@@ -7,18 +7,72 @@ import Button from './ui/Button';
...
+    this.setState({
+      ...this.state,
+      selectedCategory: value
+    });
+  };
+
+  handleRowClick = selectedQuestion => {
+    this.props.question.getQuestionById(selectedQuestion.id);
+    questionForm.set(this.props.question.singleQuestion);
+    console.log(questionForm.$('type').value);
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Remove `console.log`.

File: web/src/main/frontend/src/base/stores/QuestionStore.js (hunk 2 above)
Comment (iscekic) [reply web/src/main/frontend/src/base/stores/QuestionStore.js]: Better yet, `setActiveQuestionById` (also rename `singleQuestion` to `activeQuestion`).

File: web/src/main/frontend/src/components/Question.js [hunk 7]
```diff
@@ -7,18 +7,72 @@ import Button from './ui/Button';
...
+  };
+
   componentDidMount() {
     this.props.category.fetchCategories();
     this.props.question.fetchQuestions();
   }
 
+  handleTableSelect = selectedCategory => {
+    let value = '';
+    if (selectedCategory) {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: When is `selectedCategory` falsy?

File: web/src/main/frontend/src/components/Question.js [hunk 8]
```diff
@@ -7,18 +7,72 @@ import Button from './ui/Button';
...
+  };
+
+  handleRowClick = selectedQuestion => {
+    this.props.question.getQuestionById(selectedQuestion.id);
+    questionForm.set(this.props.question.singleQuestion);
+    console.log(questionForm.$('type').value);
+  };
+
   render() {
+    const answers = this.props.question.answers.map(answer => (
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Make this a method.

File: web/src/main/frontend/src/components/Question.js [hunk 9]
```diff
@@ -7,18 +7,72 @@ import Button from './ui/Button';
...
+              this.props.question.decrementAnswers(answer.id)
+            }
+          >
+            Remove
+          </a>
+        </div>
+      </div>
+    ));
+
+    const addAnswer = (
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Make this a method.

File: web/src/main/frontend/src/components/Question.js [hunk 10]
```diff
@@ -104,39 +168,31 @@ class Question extends Component {
...
-                  />
-                  <div className="question__controls_wrapper">
-                    <Checkbox label="Is correct" />
-                    <a className="button__link button__link-inverted">Remove</a>
-                  </div>
-                </div>
-                <div className="col-12 panel__answer_add_button">
-                  <Button cssClass="icon__block" title="+" />
-                </div>
+                {questionForm.$('type').value.value ||
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Why do we have `value.value`?

File: web/src/main/frontend/src/components/forms/QuestionForm.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  onSuccess(form) {
+    const transformedData = dataTransformer(form.values());
+    QuestionStore.submitQuestion(transformedData);
+    form.reset();
+  },
+  onError(form) {
+    form.invalidate('This is a generic error message!');
+  }
+};
+const options = {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/forms/QuestionForm.js]: `\n` above.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 12]
```diff
@@ -5,23 +5,76 @@ import questionApi from '../rest/questionApi';
...
+
   @observable questions = [];
 
-  prepareQuestionForView = q => ({text: q.text, points: q.points, id: q.id});
+  @observable answers = [{id: 1, description: '', isCorrect: ''}];
+
+  @observable selectedCategory = '';
+
+  @observable
+  questionTypes = ['FREE_FORM', 'MULTIPLE_CHOICE', 'CODING_CHALLENGE'];
```
[the remark this answers is not in the document]
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/QuestionStore.js]: Should i move it to `Question` component then?

[the remark this answers is not in the document]
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/QuestionStore.js]: Even better, make `constants.js` file, and export them as separate constants.

File: web/src/main/frontend/src/components/Question.js (hunk 7 above)
Comment (vrackovic) [reply web/src/main/frontend/src/components/Question.js]: By clicking `x` button in `Select` that is default feature of `react-select`, `selectedCategory` becomes null that resolves into app failing.

File: web/src/main/frontend/src/components/Question.js [hunk 13]
```diff
@@ -104,39 +168,31 @@ class Question extends Component {
...
+                questionForm.$('type').value === 'MULTIPLE_CHOICE' ? (
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Use constant instead of string.

File: web/src/main/frontend/src/base/constants.js [hunk 14]
```diff
@@ -0,0 +1,3 @@
+export const FREE_FORM = 'FREE_FORM';
```
Comment (iscekic) [inline web/src/main/frontend/src/base/constants.js]: Move this into `constants/questionTypes`. That way you can:

```js
import * as questionTypes from '../../constants/questionTypes';

<Select options={questionTypes} />
```

File: web/src/main/frontend/src/components/utility/stringifySnakeCase.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+const stringifySnakeCase = string => {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/utility/stringifySnakeCase.js]: No need to `export default` on separate line.

File: web/src/main/frontend/src/components/Question.js [hunk 16]
```diff
@@ -7,18 +7,71 @@ import Button from './ui/Button';
...
+    });
+  };
+
+  handleRowClick = selectedQuestion => {
+    this.props.question.setActiveQuestionById(selectedQuestion.id);
+    questionForm.set(this.props.question.activeQuestion);
+  };
+
   render() {
+    const answers = () =>
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: You misunderstood:

```js
class Question extends Component {
  getAnswers = () => {
    // ...
  }
  render() {}
}
```

File: web/src/main/frontend/src/components/Question.js (hunk 7 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/Question.js]: Exit early then:

```js
if (!selectedCategory) { return; }
```

File: web/src/main/frontend/src/components/utility/stringifySnakeCase.js [hunk 17]
```diff
@@ -0,0 +1,3 @@
+export const stringifySnakeCase = string => {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/utility/stringifySnakeCase.js]: I think this should be renamed to `underscoreToSpace`.

File: web/src/main/frontend/src/components/Question.js [hunk 18]
```diff
@@ -104,39 +169,32 @@ class Question extends Component {
...
-                  </div>
-                </div>
-                <div className="col-12 panel__answer_add_button">
-                  <Button cssClass="icon__block" title="+" />
-                </div>
+                {questionForm.$('type').value.value ||
+                questionForm.$('type').value ===
+                  questionTypes.MULTIPLE_CHOICE ? (
+                  <React.Fragment>
+                    {this.answers}
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Shouldn't these two be called - `this.answers()`?

File: web/src/main/frontend/src/components/Question.js (hunk 10 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/Question.js]: Still no answer here? 😄

Comment (vrackovic) [reply web/src/main/frontend/src/components/Question.js]: Whole flow change is needed here, will comment it and fix it after i figure out what is happening in background of mobx-react-form and react-select :laughing:

File: web/src/main/frontend/src/components/Question.js [hunk 19]
```diff
@@ -7,18 +7,98 @@ import Button from './ui/Button';
...
+    this.props.question.setActiveQuestionById(selectedQuestion.id);
+    questionForm.init(this.props.question.activeQuestion);
+    this.setState({
+      ...this.state,
+      selectedType: this.props.question.activeQuestion.type
+    });
+  };
+
+  showAnswers = () => {
+    return this.state.selectedType === questionTypes[1] ? (
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Not good. Export questions types as an object and then directly compare to the constant.

```js
export default {
  HELLO,
  WORLD
}

// elsewhere
import { HELLO } from 'constants';
```

File: web/src/main/frontend/src/components/Question.js [hunk 20]
```diff
@@ -7,18 +7,112 @@ import Button from './ui/Button';
...
+    selectedCategory: '',
+    tableSelect: ''
+  };
+
   componentDidMount() {
     this.props.category.fetchCategories();
     this.props.question.fetchQuestions();
   }
 
+  updateStateToValue(fieldName, fieldValue) {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Move everything to the question store - don't duplicate state.

File: web/src/main/frontend/src/components/ui/Select.js [hunk 21]
```diff
@@ -1,48 +1,29 @@
...
-    ));
-
+    const options = this.props.options.map(option => ({
+      value: option,
+      label: capitalize(underscoreToSpace(option.toLowerCase()))
+    }));
     return (
       <div className={this.props.containerClass}>
         <h3 className={this.props.hidden ? 'hidden__heading' : null}>
           Category:
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/Select.js]: This should be a prop - reorganize this component. 😄

File: web/src/main/frontend/src/components/Question.js [hunk 22]
```diff
@@ -4,32 +4,77 @@ import {observer, inject} from 'mobx-react';
...
+import SelectContainer from './ui/SelectContainer';
 import PanelControls from './PanelControls';
+import {questionForm} from './forms/QuestionForm';
+import questionTypes from '../base/constants/questionTypes';
+import QuestionFormAnswers from './QuestionFormAnswers';
 
 @inject('question', 'category')
 @observer
 class Question extends Component {
+  state = {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Instead of local state, let's make `tableSelect` an observable property.

File: web/src/main/frontend/src/components/Question.js [hunk 23]
```diff
@@ -4,32 +4,77 @@ import {observer, inject} from 'mobx-react';
...
+
+  handleTypeSelect = selectedType => {
+    questionForm.update({type: selectedType.value});
+  };
+
+  handleCategorySelect = selectedCategory => {
+    questionForm.update({category: selectedCategory.value});
+  };
+
+  handleTableSelect = tableSelect => {
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Looks like this method is redundant, you can just do:

```js
handleTableSelect = (option = '') => {
  // body of sortTableBySelectedOption
}
```

File: web/src/main/frontend/src/components/QuestionFormAnswers.js [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+          <Checkbox label="Is correct" />
+          <a
+            className="button__link button__link-inverted"
+            onClick={() => this.props.question.decrementAnswers(answer.id)}
+          >
+            Remove
+          </a>
+        </div>
+      </div>
+    ));
```
Comment (iscekic) [inline web/src/main/frontend/src/components/QuestionFormAnswers.js]: Missing `\n`

File: web/src/main/frontend/src/components/ui/SelectContainer.js [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+import {underscoreToSpace} from '../utility/underscoreToSpace';
+import Select from 'react-select';
+import 'react-select/dist/react-select.css';
+
+class SelectContainer extends React.Component {
+  render() {
+    const options = this.props.options.map(option => ({
+      value: option,
+      label: capitalize(underscoreToSpace(option.toLowerCase()))
+    }));
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/SelectContainer.js]: Missing `\n`

File: web/src/main/frontend/src/components/Question.js [hunk 26]
```diff
@@ -16,20 +18,49 @@ class Question extends Component {
     this.props.question.fetchQuestions();
   }
 
+  componentWillUpdate() {
+    this.showAnswers();
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Why is this here?

File: web/src/main/frontend/src/components/Question.js [hunk 27]
```diff
@@ -4,32 +4,59 @@ import {observer, inject} from 'mobx-react';
...
+
+  showAnswers = () => {
+    return this.props.questionForm.$('type').value ===
+      questionTypes.MULTIPLE_CHOICE ? (
+      <QuestionFormAnswers />
+    ) : null;
+  };
+
   render() {
+    const transformedQuestionTypes = [...Object.keys(questionTypes)];
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: Move this to `questionTypes`.

File: web/src/main/frontend/src/components/Question.js [hunk 28]
```diff
@@ -4,32 +4,59 @@ import {observer, inject} from 'mobx-react';
...
     return (
-      <div className="panel__controls">
+      <form className="panel__controls">
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Question.js]: This component is quite large - do you think it would be a good idea to split it into multiple smaller ones? @jobradovicexecom @vrackovic 

Not in this PR. 😄

File: web/src/main/frontend/src/index.js [hunk 29]
```diff
@@ -30,6 +31,7 @@ class Root extends React.Component {
           completedTest={CompletedTestStore}
           newTest={NewTestStore}
           loader={LoaderStore}
+          questionForm={questionForm}
```
Comment (iscekic) [inline web/src/main/frontend/src/index.js]: 👍 💃

File: web/src/main/frontend/src/components/Question.js (hunk 28 above)
Comment (vrackovic) [reply web/src/main/frontend/src/components/Question.js]: Yup, it should be broken into multiple peaces because it can get overwhelming even when simple things need a change

File: web/src/main/frontend/src/index.js (hunk 29 above)
Comment (vrackovic) [reply web/src/main/frontend/src/index.js]: :bowtie:
